### ISSUE TITLE
THB: knot indices can overflow

### ISSUE THREAD

**Issue by dominik-mokris-mtu** [gismo/gismo#672]
I am planning the following change. If you have any related thoughts, please let me know.

### Problem
When fitting a certain big point cloud that I am not allowed to share with a THB-surface, I started with a rather fine basis of level 0 with 127 x 15 internal knots. Fitting itself worked well but I ran into trouble when converting the resulting THB-surface to a B-spline using `gsTHBSpline<d, T>::convertToBSpline(gsTensorBSpline<d,T>& result)`: the resulting surface was severely mangled. Confusingly, using my older G+Smo installation from 2020-03-02 worked well.

After quite some debugging, the problem turned out to be one of the many changes done in #334, in particular in 53eeeb3. There the function `gsHDomain<d,T>::local2globalIndex` has been changed to use `index_t` instead of `unsigned`. On my machine, this leads to an overflow and thus to a big _negative_ value being written to `result[i]`. Instead of crashing right away, G+Smo attempts to work with this box with negative indices, leading to strange results.

### Proposed solution
After a discussion with @filiatra today, my plan is to make G+Smo (or, more precisely, `gsHTensorBasis`) to set `m_indexLevel` depending on the number of knots of the coarsest basis and the type of `index_t` to the highest value possible without overflowing.

A low-cost alternative would be to just warn the user that they are about to overflow and ask them to recompile with a bigger type used for `index_t`.

While doing so, I intend to do a little clean-up of the files in question.

### Progress
There is a dedicated branch [issue-672](https://github.com/gismo/gismo/tree/issue-672), where I try to stick to atomary commits so as not to drown in all the changes.

- [X] `gsHDomain` should consistently use its template `T`, which I will rename to `Z` in order to indicate that it is an integer type,
- [X] `kdnode` will get renamed to `gsKdNode` to be consistent with the usual G+Smo logic,
- [X] `gsAabb` will get renamed to `gsAABB` to make it clear that it is an abbreviation,
- [X] `gsVSegment` will get separated into its own file in order to be easier to locate,
- [X] commented out code will be removed,
- [X] documentation will get checked and updated if necessary,
- [X] `m_indexLevel` will be computed adaptively as suggested above,
- [X] a unit test should be added to prevent the issue from reappearing,
- [X]  fix #677 by correcting `gsTHBSpline::convertToBSpline`.

**Comment by dominik-mokris-mtu** [comment]
The includes in the `gsHSplines` folder do not exactly correspond to the dependencies between the files. This is a potential risk: the order of includes can get changed indirectly, thus leaving an innocent victim confused by many error messages.

Here's my drawing of the dependencies (I might have missed something). As there are some circular dependencies, the fix is not straightforward, I'll keep the source code as it is and leave fixing this as a nice to have ("future work^TM").

![gsHSplines-dependencies](https://github.com/gismo/gismo/assets/43567793/ece2e89b-cd2c-4a02-a8fa-0ca2510cb924)

**Comment by dominik-mokris-mtu** [comment]
Another nice to have ("future work") is to think which classes should be templated with respect to the indexing type (i.e., `Z`). However, as the hierarchy is not ordered, it is not immediatelly clear where to draw the line. Another complicating factor is that the order of templates would be different in various classes.

**Comment by dominik-mokris-mtu** [comment]
I have committed a fix for the issue together with the small code improvements listed in the description. ~~However, before a pull request we need to decide how to proceed due to the interlocking with #677 . I will talk to @filiatra and @hverhelst.~~

~~My proposal would be to merge #672 with the `m_indexLevel` computation as it is now (i.e., taking into account the strange behaviour noticed in #677 and when/if #677 gets fixed, to adapt the formula accordingly.~~

**Comment by dominik-mokris-mtu** [comment]
I have submitted the fix as a pull reques #679. I will close this issue once it is merged.
